src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +121,268 @@ namespace wasm {
+struct RefAsInfo {
+  LocalGet* target = nullptr;
+  RefAs* bestCast = nullptr;
+};
+
+// Find a cast to move earlier to another local.get. More refined subtypes are
+// chosen over less refined ones.
+struct EarlyCastFinder
+  : public LinearExecutionWalker<EarlyCastFinder,
+                                 UnifiedExpressionVisitor<EarlyCastFinder>> {
+  PassOptions options;
+  size_t numLocals;
+
+  // Tracks the current earliest local.get that we can move a cast to without
+  // side-effects, and the most refined cast that we can move to it (could be
+  // already at the earliest local.get).
+  //
+  // Note that we track a cast already on the get since we only want to move
+  // better casts to it: if the best cast is already on the get, there is no
+  // work to do.
+  std::vector<RefCastInfo> currRefCastMove;
+  std::vector<RefAsInfo> currRefAsMove;
+
+  // Used to analyze expressions to see if casts can be moved past them.
+  EffectAnalyzer testRefCast;
+  EffectAnalyzer testRefAs;
+
+  // Maps LocalGets to the most refined RefCast to move to it, to be used by the
+  // EarlyCastApplier. If the most refined RefCast is already at the desired
+  // LocalGet, it does not appear here. In the normal case, only one RefCast
+  // needs to be moved to a LocalGet. If a LocalGet is cast to multiple types
+  // which are not subtypes of each other the a trap is inevitable, and we
+  // assume this would already be optimized away beforehand, so we don't care
+  // about this special case.
+  std::unordered_map<LocalGet*, RefCast*> refCastToApply;
+
+  // Maps LocalGets to a RefAs to move to it, to be used by the
+  // EarlyCastApplier. As of right now RefAsNonNull is the only non-extern cast,
+  // so we only have one type of RefAs cast to move.
+  std::unordered_map<LocalGet*, RefAs*> refAsToApply;
+
+  EarlyCastFinder(PassOptions options, Module* module, Function* func)
+    : options(options), numLocals(func->getNumLocals()),
+      currRefCastMove(func->getNumLocals()),
+      currRefAsMove(func->getNumLocals()), testRefCast(options, *module),
+      testRefAs(options, *module) {
+
+    // TODO: generalize this when we handle more than RefAsNonNull
+    RefCast dummyRefCast(module->allocator);
+    RefAs dummyRefAs(module->allocator);
+    dummyRefAs.op = RefAsNonNull;
+
+    testRefCast.visit(&dummyRefCast);
+    testRefAs.visit(&dummyRefAs);
+  }
+
+  void flushRefCastResult(size_t index, Module& module) {
+    if (currRefCastMove[index].target) {
+      if (currRefCastMove[index].bestCast) {
+        // If the fallthrough is equal to the target, this means the cast is
+        // already at the target local.get and doesn't need to be duplicated
+        // again.
+        auto* fallthrough = Properties::getFallthrough(
+          currRefCastMove[index].bestCast, options, module);
+        if (fallthrough != currRefCastMove[index].target) {
+          refCastToApply[currRefCastMove[index].target] =
+            currRefCastMove[index].bestCast;
```

Comment:
This code can be clearer with capturing the long repeated expressions `currRefCastMove[index].target` and `currRefCastMove[index].bestCast` to locals. Using `&` on the local allows it to be modified later. I mean, this section can start with
```cpp
auto& target = currRefCastMove[index].target;
..
```

Reply:
Done.